### PR TITLE
Reportingdescriptor name to string change

### PR DESCRIPTION
**Pull request by harleenkohli**
This is a WIP. I have to add more TCs to ensure various paths to reach ReportingDescriptor.Name are tested, raising PR to get early feedback on approach.

@michaelcfanning Please review ApplyChangesFromTC32 specifically as its looking a bit complex :)

### CODE REVIEW

File: src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs [hunk 1]
```diff
@@ -135,6 +135,9 @@ private static bool ApplyChangesFromTC32(JObject sarifLog)
                 {
                     // https://github.com/oasis-tcs/sarif-spec/issues/325
                     RemoveToolLanguage(run);
+
+                    // Modify reportingDescriptor.name type to string
+                    ModifyReportingDescriptorNameType(run);
```
Comment (michaelcfanning) [inline src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs]: >ModifyReportingDescriptorNameType [](start = 20, length = 33)

ConvertReportingDescriptorNameToString

File: src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs [hunk 2]
```diff
@@ -149,6 +152,67 @@ private static void RemoveToolLanguage(JObject run)
...
+
+            // Access and modify run.conversion.tool
+            if (run["conversion"] is JObject conversion)
+            {
+                tool = (JObject)conversion["tool"];
+                ModifyReportingDescriptorNameTypeInTool(tool);
+            }
+        }
+
+        private static void ModifyReportingDescriptorNameTypeInTool(JObject tool)
```
Comment (michaelcfanning) [inline src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs]: >ModifyReportingDescriptorNameTypeInTool [](start = 28, length = 39)

ConvertToolReportingDescriptorNamesToString

File: src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs [hunk 3]
```diff
@@ -149,6 +152,67 @@ private static void RemoveToolLanguage(JObject run)
...
+        {
+            // Access and modify tool.driver
+            if (tool["driver"] is JObject driver)
+            {
+                ModifyReportingDescriptorNameTypeInToolComponent(driver);
```
Comment (michaelcfanning) [inline src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs]: >ModifyReportingDescriptorNameTypeInToolComponent [](start = 16, length = 48)

ConvertToolComponentReportingDescriptorNamesToString

File: src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs [hunk 4]
```diff
@@ -149,6 +152,67 @@ private static void RemoveToolLanguage(JObject run)
...
+            if (toolComponent["ruleDescriptors"] is JArray ruleDescriptors)
+            {
+                foreach (JObject reportingDescriptor in ruleDescriptors)
+                {
+                    ModifyReportingDescriptorNameInReportingDescriptor(reportingDescriptor);
+                }
+            }
+        }
+
+        private static void ModifyReportingDescriptorNameInReportingDescriptor(JObject reportingDescriptor)
```
Comment (michaelcfanning) [inline src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs]: >ModifyReportingDescriptorNameInReportingDescriptor [](start = 28, length = 50)

ConvertReportingDescriptorNameToString

File: src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs [hunk 5]
```diff
@@ -149,6 +152,67 @@ private static void RemoveToolLanguage(JObject run)
...
+        {
+            // Access and modify toolComponent.notificationDescriptors
+            if (toolComponent["notificationDescriptors"] is JArray notificationDescriptors)
+            {
+                foreach (JObject reportingDescriptor in notificationDescriptors)
+                {
+                    ModifyReportingDescriptorNameInReportingDescriptor(reportingDescriptor);
+                }
+            }
+
```
Comment (michaelcfanning) [inline src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs]: as discussed, push this structure into your helper to simplify this caller and expand your callee

File: src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs [hunk 6]
```diff
@@ -135,6 +135,9 @@ private static bool ApplyChangesFromTC32(JObject sarifLog)
                 {
                     // https://github.com/oasis-tcs/sarif-spec/issues/325
                     RemoveToolLanguage(run);
+
+                    // Modify reportingDescriptor.name type to string
```
Comment (harleenkohli) [inline src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs]: >// Modify reportingDescriptor.name type to string [](start = 20, length = 49)

@mikefan@microsoft.com i couldnt spot the right issue number in sarif-issues list to update here and the release notes md. Can you point to the right item?

File: src/Sarif.UnitTests/TestData/SarifVersionOneToCurrentVisitor/ExpectedOutputs/OneRunWithAllReportingDescriptors.sarif [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+          "semanticVersion": "2.1.0",
+          "ruleDescriptors": [
+            {
+              "id": "C2003",
+              "name": "Rule C2003",
+              "shortDescription": {
+                "text": "Rules were meant to be broken."
+              },
+              "fullDescription": {
+                "text": "This is a test for validating modification of tool.driver.ruleDescriptors"
```
Comment (harleenkohli) [inline src/Sarif.UnitTests/TestData/SarifVersionOneToCurrentVisitor/ExpectedOutputs/OneRunWithAllReportingDescriptors.sarif]: >tool.driver.ruleDescriptors [](start = 71, length = 27)

i might need a bit of help in coming up with more TCs to ensure all paths are tested. Will sync up offline. I am ok adding those as part of separate PR is required.

File: src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs (hunk 5 above)
Comment (harleenkohli) [reply src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs]: i kept the if condition in the caller

---
In reply to: [260416838](https://github.com/Microsoft/sarif-sdk/pull/1309#discussion_r260416838) [](ancestors = 260416838)

File: src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs (hunk 6 above)
Comment (michaelcfanning) [reply src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs]: These changes are part of the localization removal issue, see below.

[https://github.com/oasis-tcs/sarif-spec/issues/325](https://github.com/oasis-tcs/sarif-spec/issues/325)

---
In reply to: [260425123](https://github.com/Microsoft/sarif-sdk/pull/1309#discussion_r260425123) [](ancestors = 260425123)

File: src/Sarif.UnitTests/TestData/SarifVersionOneToCurrentVisitor/ExpectedOutputs/OneRunWithAllReportingDescriptors.sarif [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+              },
+              "defaultConfiguration": {
+                "enabled": false,
+                "level": "note",
+                "rank": -1.0
```
Comment (michaelcfanning) [inline src/Sarif.UnitTests/TestData/SarifVersionOneToCurrentVisitor/ExpectedOutputs/OneRunWithAllReportingDescriptors.sarif]: this is wrong, this is a default value. Did you lose a custom edit to a 'NotYetGeneratedFiles'? Check the default value of the rank attribute, which s/be expressed as -1.0 (not 1.0).

Underlying issue is that if you emit -1 as a default value, this looks like an integer, so the actual default of -1.0 (which is a different JSON numeric type) will be explicit in code.

**Comment by michaelcfanning** [comment]
>                 "rank": -1.0

remove all default ranks of -1.0. they should always be elided from persisted content. this content isn't strictly illegal per schema, of course. 

---
Refers to: src/Sarif.UnitTests/TestData/SarifVersionOneToCurrentVisitor/ExpectedOutputs/OneRunWithRules.sarif:32 in 04763a1. [](commit_id = 04763a12e755622340265f1ec9469a1a81c7efe7, deletion_comment = False)

File: src/Sarif.UnitTests/TestData/SarifVersionOneToCurrentVisitor/Inputs/OneRunWithAllReportingDescriptors.sarif [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  "runs": [
+    {
+      "rules": {
+        "C2003": {
+          "id": "C2003",
+          "name": "Rule C2003",
+          "shortDescription": "Rules were meant to be broken.",
+          "fullDescription": "This is a test for validating modification of tool.driver.ruleDescriptors",
+          "configuration": "disabled",
+          "defaultLevel": "pass"
```
Comment (michaelcfanning) [inline src/Sarif.UnitTests/TestData/SarifVersionOneToCurrentVisitor/Inputs/OneRunWithAllReportingDescriptors.sarif]: >"pass" [](start = 26, length = 6)

a default level of pass? this is wrong, it s/be a failure level. did you get this pattern from an existing test?

File: src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs (hunk 6 above)
Comment (harleenkohli) [reply src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs]: thanks! will update

---
In reply to: [260549244](https://github.com/Microsoft/sarif-sdk/pull/1309#discussion_r260549244) [](ancestors = 260549244,260425123)

File: src/Sarif.UnitTests/TestData/SarifVersionOneToCurrentVisitor/Inputs/OneRunWithAllReportingDescriptors.sarif (hunk 9 above)
Comment (harleenkohli) [reply src/Sarif.UnitTests/TestData/SarifVersionOneToCurrentVisitor/Inputs/OneRunWithAllReportingDescriptors.sarif]: yes, let me fix it in source as well!

---
In reply to: [260550146](https://github.com/Microsoft/sarif-sdk/pull/1309#discussion_r260550146) [](ancestors = 260550146)

File: src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs (hunk 6 above)
Comment (harleenkohli) [reply src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs]: we have an entry for this in release notes, thanks much!

---
In reply to: [260551276](https://github.com/Microsoft/sarif-sdk/pull/1309#discussion_r260551276) [](ancestors = 260551276,260549244,260425123)

**Comment by harleenkohli** [comment]
@michaelcfanning we resolved the discussed comments. If there are no additional comments, i love for this one to get merged before i start the final changes for localization removal. Can you please take a look :-)

**Pull request merged**